volttrontesting/utils/platformwrapper.py

```diff
@@ -1001,21 +1071,52 @@ def install_agent(self, agent_wheel=None, agent_dir=None, config_file=None,
 
         return agent_uuid
 
+    def __wait_for_control_connection_to_exit__(self, timeout: int = 20):
+        """
+        Call the dynamic agent's peerlist method until the control connection is no longer connected or
+        timeout is reached
+        :param timeout:
+        :return:
+        """
+        self.logit("Waiting for control_connection to exit")
+        disconnected = False
+        timer_start = time.time()
+        while not disconnected:
+            peers = self.dynamic_agent.vip.peerlist().get(timeout=20)
+            disconnected = CONTROL_CONNECTION not in peers
+            if disconnected:
+                break
+            self.logit(f"Waiting for control connection to disconnect: {peers} time: {timer_start - time.time()} timeout is {timeout}")
+            if time.time() - timer_start > timeout:
+                raise PlatformWrapperError(f"Failed for {CONTROL_CONNECTION} to exit in a timely manner.")
+            time.sleep(0.5)
+
+        if not disconnected:
+            raise PlatformWrapperError("Control connection did not stop properly")
+
     def start_agent(self, agent_uuid):
         self.logit('Starting agent {}'.format(agent_uuid))
         self.logit("VOLTTRON_HOME SETTING: {}".format(
             self.env['VOLTTRON_HOME']))
-        cmd = ['volttron-ctl', '--json']
-        cmd.extend(['start', agent_uuid])
+        if not self.is_running():
+            raise PlatformWrapperError("Instance must be running before starting agent")
 
+        self.__wait_for_control_connection_to_exit__(5)
+
+        cmd = [self.vctl_exe, '--json']
+        cmd.extend(['start', agent_uuid])
         result = execute_command(cmd, self.env)
 
+        self.__wait_for_control_connection_to_exit__(5)
```

Comment:
My understanding is that we are waiting here to make sure vctl start has completed all its work and relinquished its connection. But why should we wait before executing vctl start? Shouldn't we just wait after a vctl operations - start, stop, remove ? Do we need a wait after vctl status ?

Reply:
We need to wait for any calls that we shell out to vctl...each time we do that we have to wait until the previous exits.  that's why.  In general when running from command line it's not an issue but for this it is.  Besides having the wait here proves that we are ready to continue with execution.